samza-core/src/main/java/org/apache/samza/operators/impl/store/TimeSeriesStore.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package org.apache.samza.operators.impl.store;
+
+import org.apache.samza.storage.kv.ClosableIterator;
+
+/**
+ * A key-value store that allows entries to be queried and stored based on time ranges.
+ *
+ * Operations on the store can be invoked from multiple threads. Hence, implementations are expected to be thread-safe.
+ *
+ * @param <K> the type of key in the store
+ * @param <V> the type of value in the store
+ */
+
+public interface TimeSeriesStore<K, V> {
+
+  /**
+   * Insert a key and the value to the store with the provided timestamp.
+   *
+   * @param key the key to insert
+   * @param val the value to insert
+   * @param timeStamp the timestamp in milliseconds
+   */
+  void put(K key, V val, Long timeStamp);
+
+  /**
+   * Returns an iterator over values for the given key in the provided time-range - [{@code startTimestamp}, {@code endTimestamp})
+   *
+   * Values returned by the iterator are ordered by their timestamp. Values with the same timestamp are
+   * returned in their order of insertion.
+   *
+   * @param key the key to look up in the store
+   * @param startTimestamp the start timestamp of the range, inclusive
+   * @param endTimeStamp the end timestamp of the range, exclusive
+   * @throws IllegalArgumentException when startTimeStamp &gt; endTimeStamp, or when either of them is negative
+   */
+  ClosableIterator<TimeSeriesValue<V>> get(K key, Long startTimestamp, Long endTimeStamp);
```

Comment:
Do we need range(K k1, K k2)/all?

Reply:
There's no usecase currently that requires a range scan across keys, or an iterator across all keys in the store. We agreed that we'll add it should a scenario arise